exonum/src/messages2/protocol.rs

```diff
@@ -0,0 +1,709 @@
+    }
+}
+
+/// Full message constraints list.
+#[doc(hidden)]
+pub trait ProtocolMessage: Debug + Clone + BinaryForm {
+    fn message_type() -> (u8, u8);
+    ///Trying to convert `Protocol` to concrete message,
+    ///if ok returns message `Message<Self>` if fails, returns `Protocol` back.
+    fn try_from(p: Protocol) -> Result<Message<Self>, Protocol>;
+
+    fn into_protocol(this: Message<Self>) -> Protocol;
+
+    fn into_message_from_parts(self, sm: SignedMessage) -> Message<Self>;
+}
+
+/// Implement Exonum message protocol.
+///
+/// Protocol should be described according to format:
+/// ```
+/// /// type of SignedMessage => new name of Protocol enum.
+/// SignedMessage => Protocol {
+///       // class ID => class name
+///       0 => Service {
+///            // message = message type ID
+///            RawTransaction = 0,
+///            Connect = 1,
+///            Status = 2,
+///            // ...
+///        },п
+///        1 => Consensus {
+///            Precommit = 0,
+///            Propose = 1,
+///            Prevote = 2,
+///        },
+/// }
+/// ```
+/// where:
+/// `$ProtocolName` is an name of protocol which is described by this schema,
+///     enum with same name will be created. All message classes are encapsulate by this enum;
+/// `$MessageClass` is a module name which is designed to handle messages,
+///     enum with same name will be crated. All message within `MessageClass` should be unique;
+/// `$MessageType` is a concrete messages within some `$MessageClass`;
+/// `$SignedMessage` is a typename of `SignedMessage` from which $ProtocolName could be created;
+/// `$cls_num` and `$typ_num` is a constant which represent
+///     message class and message type respectively
+///
+/// Each `$MessageType` should implement `Clone` and `Debug`.
+///
+macro_rules! impl_protocol {
+
+    (
+    $(#[$attr:meta])+
+    $signed_message:ident => $protocol_name:ident{
+        $($(#[$attr_class:meta])+
+        $class_num:expr => $class:ident{
+            $(
+                $(#[$attr_type:meta])+
+                $type:ident = $type_num:expr
+            ),+ $(,)*
+        } $(,)*)+
+    }
+    ) => {
+
+        $(
+            #[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Clone)]
+            $(#[$attr_class])+
+            pub enum $class {
+            $(
+                $(#[$attr_type])+
+                $type(Message<$type>)
+            ),+
+            }
+
+            $(
+
+            impl ProtocolMessage for $type {
+                fn message_type() -> (u8, u8) {
+                    ($class_num, $type_num)
+                }
+
+                fn try_from(p: $protocol_name) -> Result<Message<Self>,Protocol> {
+                    match p {
+                        $protocol_name::$class($class::$type(s)) => Ok(s),
+                        p => Err(p)
+                    }
+                }
+
+                fn into_protocol(this: Message<Self>) -> Protocol {
```

Comment:
Name sounds a little bit unclear. Why not just to implement `From` conversion?

Reply:
Why it's not clear to you?
This is an internal trait that provide all kind function for being a part of `Protocol`.

Without it we should everywhere add bound

`fn bla<T: ProtocolMessage>(t:T)`
vs
`fn bla<T>(t:T)
where T:ProtocolMessage,
Message<T>: Into<Protocol>
`

which probably is not clear.

Additionally, `Into` can be implemented to multiple type params, this mean you should explicitly define to what type you want to convert.

```
let t: Message<X> = ...
foo(t.into());
```